contracts/actions/uniswap/v3/UniCreatePoolV3.sol

```diff
@@ -0,0 +1,166 @@
+        bytes32[] memory _returnValues
+    ) public payable virtual override returns (bytes32) {
+        Params memory inputData = parseInputs(_callData);
+
+        inputData.amount0Desired = _parseParamUint(
+            inputData.amount0Desired,
+            _paramMapping[0],
+            _subData,
+            _returnValues
+        );
+        inputData.amount1Desired = _parseParamUint(
+            inputData.amount1Desired,
+            _paramMapping[1],
+            _subData,
+            _returnValues
+        );
+
+        _createPool(inputData);
+
+        uint256 tokenId = _uniCreatePosition(inputData);
+
+        return bytes32(tokenId);
+    }
+
+    /// @inheritdoc ActionBase
+    function executeActionDirect(bytes[] memory _callData) public payable override {
+        Params memory inputData = parseInputs(_callData);
+
+        _createPool(inputData);
+        _uniCreatePosition(inputData);
+    }
+
+    /// @inheritdoc ActionBase
+    function actionType() public pure virtual override returns (uint8) {
+        return uint8(ActionType.STANDARD_ACTION);
+    }
+
+    //////////////////////////// ACTION LOGIC ////////////////////////////
+
+    function _createPool(Params memory _inputData) internal returns (address pool) {
+        pool = positionManager.createAndInitializePoolIfNecessary(
+            _inputData.token0,
+            _inputData.token1,
+            _inputData.fee,
+            _inputData.sqrtPriceX96
+        );
+    }
+
+    function _uniCreatePosition(Params memory _inputData) internal returns (uint256 tokenId) {
+        // fetch tokens from address;
+        uint256 amount0Pulled = _inputData.token0.pullTokensIfNeeded(
+            _inputData.from,
+            _inputData.amount0Desired
+        );
+        uint256 amount1Pulled = _inputData.token1.pullTokensIfNeeded(
+            _inputData.from,
+            _inputData.amount1Desired
+        );
+
+        // approve positionManager so it can pull tokens
+        _inputData.token0.approveToken(address(positionManager), amount0Pulled);
+        _inputData.token1.approveToken(address(positionManager), amount1Pulled);
+
+        _inputData.amount0Desired = amount0Pulled;
+        _inputData.amount1Desired = amount1Pulled;
+
+        uint128 liquidity;
+        uint256 amount0;
+        uint256 amount1;
+        (tokenId, liquidity, amount0, amount1) = _uniMint(_inputData);
+
+        //send leftovers
+        _inputData.token0.withdrawTokens(_inputData.from, sub(_inputData.amount0Desired, amount0));
+        _inputData.token1.withdrawTokens(_inputData.from, sub(_inputData.amount1Desired, amount1));
+
+        logger.Log(
+            address(this),
+            msg.sender,
+            "UniCreatePoolV3",
+            abi.encode(_inputData, tokenId, liquidity, amount0, amount1)
+        );
+    }
+
+    /// @dev mints new NFT that represents a position with selected parameters
+    /// @return tokenId of new NFT, how much liquidity it now has and token amounts
```

Comment:
Describe what are token amounts, is it the leftovers or is it how much is actually put in the pool